src/backend/access/transam/xlog.c

```diff
@@ -9366,3 +9366,26 @@ SetWalWriterSleeping(bool sleeping)
 	XLogCtl->WalWriterSleeping = sleeping;
 	SpinLockRelease(&XLogCtl->info_lck);
 }
+
+void
+XLogUpdateWalBuffers(char* data, XLogRecPtr start, size_t len)
+{
+	XLogRecPtr end;
+	int idx;
+	XLogRecPtr pagebegptr;
+
+	LWLockAcquire(WALBufMappingLock, LW_EXCLUSIVE);
+
+	end = start + len;
+	idx = XLogRecPtrToBufIdx(end);
+	pagebegptr = XLogCtl->xlblocks[idx] - XLOG_BLCKSZ;
```

Comment:
This page might be already written out, in which case we shouldn't override it here.

Reply:
What's wrong happen if we update already written page?
Certainly it is possible to add some check here (compare `end` with `XLogCtl->LogwrtRqst.Write`) but I do not see any reason for it. We copy at most one page and only in case of walproposer recovery. So it should not somehow affect performance.

Or do you think that it can somehow violate consistency of way buffers?

Reply:
> What's wrong happen if we update already written page?

The slot will be occupied by another probably not yet written out page.

Reply:
In this case correspondent `xlblocks` entry should be updated, should not it?

Reply:
Ah yes, it seems so, not deeply familiar with this.